modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStatsResults.java

```diff
@@ -0,0 +1,223 @@
+
+    /** Base ctor */
+    private MatrixStatsResults() {
+        results = RunningStats.EMPTY();
+        this.correlation = new HashMap<>();
+    }
+
+    /** creates and computes result from provided stats */
+    public MatrixStatsResults(RunningStats stats) {
+        this.results = stats.clone();
+        this.correlation = new HashMap<>();
+        this.compute();
+    }
+
+    /** creates a results object from the given stream */
+    @SuppressWarnings("unchecked")
+    protected MatrixStatsResults(StreamInput in) {
+        try {
+            results = new RunningStats(in);
+            correlation = (HashMap<String, HashMap<String, Double>>) in.readGenericValue();
+        } catch (IOException e) {
+            throw new ElasticsearchException("Error trying to create multifield_stats results from stream input", e);
+        }
+    }
+
+    /** Marshalls MatrixStatsResults */
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        // marshall results
+        results.writeTo(out);
+        // marshall correlation
+        out.writeGenericValue(correlation);
+    }
+
+    /** create an empty results object **/
+    protected static MatrixStatsResults EMPTY() {
+        return new MatrixStatsResults();
+    }
+
+    /** return document count */
+    public final long getDocCount() {
+        return results.docCount;
+    }
+
+    /** return the field counts */
+    public Map<String, Long> getFieldCounts() {
+        return Collections.unmodifiableMap(results.counts);
+    }
+
+    /** return the fied count for the requested field */
+    public long getFieldCount(String field) {
+        if (results.counts.containsKey(field) == false) {
+            return 0;
+        }
+        return results.counts.get(field);
+    }
+
+    /** return the means */
+    public Map<String, Double> getMeans() {
+        return Collections.unmodifiableMap(results.means);
+    }
+
+    /** return the mean for the requested field */
+    public Double getMean(String field) {
+        return results.means.get(field);
+    }
+
+    /** return the variances */
+    public Map<String, Double> getVariances() {
+        return Collections.unmodifiableMap(results.variances);
+    }
+
+    /** return the variance for the requested field */
+    public Double getVariance(String field) {
+        return results.variances.get(field);
+    }
+
+    /** return the skewness */
+    public Map<String, Double> getSkewness() {
+        return Collections.unmodifiableMap(results.skewness);
+    }
+
+    /** return the skewness for the requested field */
+    public Double getSkewness(String field) {
+        return results.skewness.get(field);
+    }
+
+    /** return the kurtosis */
+    public Map<String, Double> getKurtosis() {
+        return Collections.unmodifiableMap(results.kurtosis);
+    }
+
+    /** return the kurtosis for the requested field */
+    public Double getKurtosis(String field) {
```

Comment:
should it and other methods in this class return double rather than Double (and probably thorw an exception if a field is requested that was not specified at request time)?